File: 2192-all-ancestors-of-a-node-in-a-directed-acyclic-graph/2192-all-ancestors-of-a-node-in-a-directed-acyclic-graph.py
class Solution:
    def getAncestors(self, n: int, edges: List[List[int]]) -> List[List[int]]:
        adj = defaultdict(list)
        indegree = [0] * n
        for u, v in edges:
            adj[u].append(v)
            indegree[v] += 1
            
        q = deque([i for i in range(n) if indegree[i] == 0])
        ancestors = [set() for _ in range(n)]
        
        while q:
            curr = q.popleft()
            for neighbor in adj[curr]:
                ancestors[neighbor].add(curr)
                ancestors[neighbor].update(ancestors[curr])
                indegree[neighbor] -= 1
                if indegree[neighbor] == 0:
                    q.append(neighbor)
        
        return [sorted(list(ancestor_set)) for ancestor_set in ancestors]






# class Solution:
#     def solve(self, mp, node, temp, visited):
#         if visited[node]:
#             return
#         visited[node] = True
        
#         for ancestor in mp[node]:
#             if not visited[ancestor]:
#                 temp.append(ancestor)
#                 self.solve(mp, ancestor, temp, visited)

#     def getAncestors(self, n: int, edges: List[List[int]]) -> List[List[int]]:
#         mp = defaultdict(list)
#         ans = [[] for _ in range(n)]
        
#         for edge in edges:
#             mp[edge[1]].append(edge[0])

#         for i in range(n):
#             temp = []
#             visited = [False] * n
#             self.solve(mp, i, temp, visited)
#             temp.sort()
#             ans[i] = temp

#         return ans

File: LeetCode/2192-all-ancestors-of-a-node-in-a-directed-acyclic-graph/2192-all-ancestors-of-a-node-in-a-directed-acyclic-graph.java
class Solution {
    public List<List<Integer>> getAncestors(int n, int[][] edges) {
        List<Integer>[] adj = new ArrayList[n];
        int[] indegree = new int[n];
        for (int i = 0; i < n; i++) {
            adj[i] = new ArrayList<>();
        }
        
        for (int[] edge : edges) {
            adj[edge[0]].add(edge[1]);
            indegree[edge[1]]++;
        }
        
        Queue<Integer> q = new LinkedList<>();
        for (int i = 0; i < n; i++) {
            if (indegree[i] == 0) {
                q.add(i);
            }
        }
        
        Set<Integer>[] ancestors = new HashSet[n];
        for (int i = 0; i < n; i++) {
            ancestors[i] = new HashSet<>();
        }
        
        while (!q.isEmpty()) {
            int curr = q.poll();
            for (int neighbor : adj[curr]) {
                ancestors[neighbor].add(curr);
                ancestors[neighbor].addAll(ancestors[curr]);
                indegree[neighbor]--;
                if (indegree[neighbor] == 0) {
                    q.add(neighbor);
                }
            }
        }
        
        List<List<Integer>> ans = new ArrayList<>();
        for (int i = 0; i < n; i++) {
            List<Integer> list = new ArrayList<>(ancestors[i]);
            Collections.sort(list);
            ans.add(list);
        }
        
        return ans;
    }
}





// class Solution {
//     public void solve(Map<Integer, List<Integer>> mp, int node, List<Integer> temp, boolean[] vis) {
//         if (vis[node]) {
//             return;
//         }
//         vis[node] = true;
//         if (mp.containsKey(node)) {
//             for (int x : mp.get(node)) {
//                 if (!vis[x]) {
//                     temp.add(x);
//                     solve(mp, x, temp, vis);
//                 }
//             }
//         }
//     }

//     public List<List<Integer>> getAncestors(int n, int[][] edges) {
//         List<List<Integer>> ans = new ArrayList<>(n);
//         Map<Integer, List<Integer>> mp = new HashMap<>();

//         for (int i = 0; i < n; i++) {
//             ans.add(new ArrayList<>());
//         }

//         for (int[] edge : edges) {
//             mp.computeIfAbsent(edge[1], k -> new ArrayList<>()).add(edge[0]);
//         }

//         for (int i = 0; i < n; i++) {
//             List<Integer> temp = new ArrayList<>();
//             boolean[] vis = new boolean[n];
//             solve(mp, i, temp, vis);
//             Collections.sort(temp);
//             ans.set(i, temp);
//         }

//         return ans;
//     }
// }